crates/shared/src/price_estimation/cached.rs

```diff
@@ -0,0 +1,107 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use cached::{Cached, TimedSizedCache};
+use std::{sync::Mutex, time::Duration};
```

Comment:
Don't we have to use `tokio::sync::Mutex` to avoid stopping every future in the current thread?
I believe if we make the `CachingPriceEstimator` the outer most layer of the estimator hierarchy this is probably fine but if we were to only wrap the "leaf" estimators in this cache this would lead to problems, no?

Reply:
I think `tokio::sync::Mutex` is only needed if the lock is held across `await`s. Otherwise, I imagine `std::sync::Mutex` to actually be more performant for this CPU bound code.

From the [`tokio-rs` docs](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) itself:
> Contrary to popular belief, it is ok and often preferred to use the ordinary Mutex from the standard library in asynchronous code.

Reply:
Also, the compiler helps you out a bit here if you try to make a `BoxFuture` where you hold `MutexGuard`s across `await`s:
```rust
use futures::future::{BoxFuture, FutureExt as _};
use std::{sync::Mutex, future::Future};

fn foo(thing: impl Future<Output = ()> + Send + 'static) -> BoxFuture<'static, ()> {
    let mutex = Mutex::new(1u64);
    async move {
        // thing.await; // Here is OK.
        {
            let mut guard = mutex.lock().unwrap();
            thing.await; // Here is NOT OK.
            *guard += 1;
        }
        // thing.await; // Here is OK.
    }.boxed()
}
```

Will error:
```
error: future cannot be sent between threads safely
  --> src/main.rs:16:7
   |
16 |     }.boxed()
   |       ^^^^^ future created by async block is not `Send`
   |
   = help: within `impl futures::Future`, the trait `std::marker::Send` is not implemented for `std::sync::MutexGuard<'_, u64>`
note: future is not `Send` as this value is used across an await
  --> src/main.rs:12:13
   |
11 |             let mut guard = mutex.lock().unwrap();
   |                 --------- has type `std::sync::MutexGuard<'_, u64>` which is not `Send`
12 |             thing.await; // Here is NOT OK.
   |             ^^^^^^^^^^^ await occurs here, with `mut guard` maybe used later
13 |             *guard += 1;
14 |         }
   |         - `mut guard` is later dropped here
```

Playground link: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7af6ba330b2b4cb2211c748411a749fa>

Reply:
You have to use an async aware mutex when you hold the lock across awaits or get deadlocks. When it isn't held across awaits like in this case you can decide. The mutex is only locked for short duration (read or write a couple of elements into the cache) so it will not block async executions for a long time. Non async mutexes have overall less overhead so I use them in situations like these.

Reply:
Thanks for the explanation. It seems like I have to do more research.
Then the PR seems fine. Will approve.

Reply:
The [shared state](https://tokio.rs/tokio/tutorial/shared-state) section of the tokio tutorial also has some great insights on this topic. It specifically addresses some edge cases I was still unsure about. Namely high contention and it seems `tokio::sync::Mutex` wouldn't help in that scenario either.